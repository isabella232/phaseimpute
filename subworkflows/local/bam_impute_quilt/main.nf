include { QUILT_QUILT                        } from '../../../modules/nf-core/quilt/quilt'
include { BCFTOOLS_ANNOTATE                  } from '../../../modules/nf-core/bcftools/annotate'

workflow BAM_IMPUTE_QUILT {

    take:
    ch_input             // channel: [ [id], [bam], [bai], bamlist ]
    ch_hap_legend        // channel: [ [panel, chr], hap, legend ]
    ch_chunks            // channel: [ [panel, chr], chr, start_coordinate, end_coordinate ]
    ch_fasta             // channel: [ [genome], fa, fai ]

    main:

    ch_versions = Channel.empty()

    posfile             = []
    phasefile           = []
    posfile_phasefile   = [[id: null], posfile, phasefile]
    genetic_map_file    = []

    ngen_params         = params.ngen
    buffer_params       = params.buffer

    ch_hap_chunks = ch_hap_legend
        .combine(ch_chunks, by:0)
        .map { it + ngen_params + buffer_params + [[]] }

    if (!genetic_map_file.isEmpty()) {
        // Add genetic map file (untested)
        ch_hap_chunks = ch_hap_chunks
            .map{it[0..-1]}
            .join(genetic_map_file)
    }

    ch_quilt = ch_input
        .combine(ch_hap_chunks)
        .map {
            metaI, bam, bai, bamlist, metaPC, hap, legend, chr, start, end, ngen, buffer, gmap ->
            [
                metaI + [panel: metaPC.id, chr: metaPC.chr, chunk: metaPC.chr + ":" + start + "-" + end],
                bam, bai, bamlist, hap, legend, chr, start, end, ngen, buffer, gmap
            ]
        }

    // Run QUILT
    QUILT_QUILT ( ch_quilt, posfile_phasefile, ch_fasta )
    ch_versions = ch_versions.mix(QUILT_QUILT.out.versions.first())

    // Annotate the variants
    BCFTOOLS_ANNOTATE(QUILT_QUILT.out.vcf
        .join(QUILT_QUILT.out.tbi)
        .combine(Channel.of([[], [], [], []]))
    )
    ch_versions = ch_versions.mix(BCFTOOLS_ANNOTATE.out.versions.first())

    // Join VCFs and TBIs
    ch_vcf_tbi = BCFTOOLS_ANNOTATE.out.vcf
        .join(BCFTOOLS_ANNOTATE.out.tbi)
        .map { metaIPC, vcf, tbi -> [metaIPC + [tools: "quilt"], vcf, tbi] }

    emit:
    vcf_tbi     = ch_vcf_tbi               // channel:  [ [id, panel], vcf, tbi ]
    versions    = ch_versions              // channel:  [ versions.yml ]
}
