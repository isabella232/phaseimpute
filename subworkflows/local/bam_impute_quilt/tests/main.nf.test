nextflow_workflow {

    name "Test Subworkflow BAM_IMPUTE_QUILT"
    script "../main.nf"
    config "./nextflow.config"

    workflow "BAM_IMPUTE_QUILT"

    tag "subworkflows"
    tag "subworkflows_local"
    tag "subworkflows/bam_impute_quilt"
    tag "bam_impute_quilt"

    tag "quilt"
    tag "quilt/quilt"
    tag "bcftools"
    tag "bcftools/index"
    tag "bcftools/annotate"

    test("Impute with quilt two individuals with bamlist") {
        when {
            workflow {
                """
                bamlist = Channel.of(
                    "NA12878.s.bam",
                    "NA19401.s.bam"
                ).collectFile(name: 'bamlist.txt', newLine: true)
                input[0] = Channel.fromList([
                    [
                        [id: "allid"],
                        [
                            file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s.bam", checkIfExist:true),
                            file(params.pipelines_testdata_base_path + "hum_data/individuals/NA19401/NA19401.s.bam", checkIfExist:true)
                        ],
                        [
                            file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s.bam.bai", checkIfExist:true),
                            file(params.pipelines_testdata_base_path + "hum_data/individuals/NA19401/NA19401.s.bam.bai", checkIfExist:true)
                        ]
                    ],
                ]).combine(bamlist)
                input[1] = Channel.of([
                        [id: "1000GP", chr: "22"],
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.hap.gz", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.legend.gz", checkIfExist:true)
                    ]).collect()
                input[2] = Channel.of(
                    [[chr: "22", id: "1000GP"], "chr22", "16570065", "16597215"],
                    [[chr: "22", id: "1000GP"], "chr22", "16587172", "16609999"]
                )
                input[3] = Channel.of([
                    [id: "GRCh38"],
                    file(params.pipelines_testdata_base_path + "hum_data/reference_genome/GRCh38.s.fa.gz", checkIfExist:true)
                ]).collect()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.versions,
                    workflow.out.vcf_tbi.collect{[
                        it[0],
                        path(it[1]).getFileName().toString(),
                        path(it[2]).getFileName().toString()
                    ] },
                    workflow.out.vcf_tbi.collect{ path(it[1]).vcf.summary },
                    workflow.out.vcf_tbi.collect{ path(it[1]).vcf.header.getGenotypeSamples().sort() }
                ).match() }
            )
        }
    }

    test("Impute with quilt one individuals") {
        when {
            workflow {
                """
                input[0] = Channel.fromList([
                    [
                        [id: "NA12878"],
                        file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s.bam", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s.bam.bai", checkIfExist:true),
                        []
                    ]
                ])
                input[1] = Channel.of([
                        [id: "1000GP", chr: "22"],
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.hap.gz", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.legend.gz", checkIfExist:true)
                    ]).collect()
                input[2] = Channel.of(
                    [[chr: "22", id: "1000GP"], "chr22", "16570065", "16597215"],
                    [[chr: "22", id: "1000GP"], "chr22", "16587172", "16609999"]
                )
                input[3] = Channel.of([
                    [id: "GRCh38"],
                    file(params.pipelines_testdata_base_path + "hum_data/reference_genome/GRCh38.s.fa.gz", checkIfExist:true)
                ]).collect()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.versions,
                    workflow.out.vcf_tbi.collect{[
                        it[0],
                        path(it[1]).getFileName().toString(),
                        path(it[2]).getFileName().toString()
                    ] },
                    workflow.out.vcf_tbi.collect{ path(it[1]).vcf.summary },
                    workflow.out.vcf_tbi.collect{ path(it[1]).vcf.header.getGenotypeSamples().sort() }
                ).match() }
            )
        }
    }
}
