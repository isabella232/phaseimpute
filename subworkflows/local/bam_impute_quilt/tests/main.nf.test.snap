{
    "Impute with quilt two individuals with bamlist": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "allid",
                            "panel": "1000GP",
                            "chr": "22",
                            "chunk": "16570065-16597215",
                            "tools": "quilt"
                        },
                        "allid_22_16570065-16597215.impute.annotate.vcf.gz:md5,de611972f693a8c90c4a8175deab5249",
                        "allid_22_16570065-16597215.impute.annotate.vcf.gz.tbi:md5,0802cd70847696ee407c520fd963db31"
                    ],
                    [
                        {
                            "id": "allid",
                            "panel": "1000GP",
                            "chr": "22",
                            "chunk": "16587172-16609999",
                            "tools": "quilt"
                        },
                        "allid_22_16587172-16609999.impute.annotate.vcf.gz:md5,8a1dc994fc0eb4771b6acfe0e876580b",
                        "allid_22_16587172-16609999.impute.annotate.vcf.gz.tbi:md5,f3e0d1c94d02bcb57bafd143c49269f4"
                    ]
                ],
                "1": [
                    "versions.yml:md5,ac2d0f2a8e77cd796d3b4b84fdd46b0d",
                    "versions.yml:md5,cb038a1f67a47a76d0e0a983cfd56789"
                ],
                "vcf_tbi": [
                    [
                        {
                            "id": "allid",
                            "panel": "1000GP",
                            "chr": "22",
                            "chunk": "16570065-16597215",
                            "tools": "quilt"
                        },
                        "allid_22_16570065-16597215.impute.annotate.vcf.gz:md5,de611972f693a8c90c4a8175deab5249",
                        "allid_22_16570065-16597215.impute.annotate.vcf.gz.tbi:md5,0802cd70847696ee407c520fd963db31"
                    ],
                    [
                        {
                            "id": "allid",
                            "panel": "1000GP",
                            "chr": "22",
                            "chunk": "16587172-16609999",
                            "tools": "quilt"
                        },
                        "allid_22_16587172-16609999.impute.annotate.vcf.gz:md5,8a1dc994fc0eb4771b6acfe0e876580b",
                        "allid_22_16587172-16609999.impute.annotate.vcf.gz.tbi:md5,f3e0d1c94d02bcb57bafd143c49269f4"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,ac2d0f2a8e77cd796d3b4b84fdd46b0d",
                    "versions.yml:md5,cb038a1f67a47a76d0e0a983cfd56789"
                ]
            },
            [
                "VcfFile [chromosomes=[chr22], sampleCount=2, variantCount=572, phased=true, phasedAutodetect=false]",
                "VcfFile [chromosomes=[chr22], sampleCount=2, variantCount=570, phased=true, phasedAutodetect=false]"
            ],
            [
                [
                    "NA12878",
                    "NA19401"
                ],
                [
                    "NA12878",
                    "NA19401"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-11-06T16:24:00.161098553"
    },
    "Impute with quilt one individuals": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "NA12878",
                            "panel": "1000GP",
                            "chr": "22",
                            "chunk": "16570065-16597215",
                            "tools": "quilt"
                        },
                        "NA12878_22_16570065-16597215.impute.annotate.vcf.gz:md5,f1a6f74a4782fc1df96db2b1fd458381",
                        "NA12878_22_16570065-16597215.impute.annotate.vcf.gz.tbi:md5,de3b20dc5f67aae2a9669f0c667edb5d"
                    ],
                    [
                        {
                            "id": "NA12878",
                            "panel": "1000GP",
                            "chr": "22",
                            "chunk": "16587172-16609999",
                            "tools": "quilt"
                        },
                        "NA12878_22_16587172-16609999.impute.annotate.vcf.gz:md5,d549aa4a26cbe659081002521c4400c2",
                        "NA12878_22_16587172-16609999.impute.annotate.vcf.gz.tbi:md5,674782f71cca76015e0b72f9bc4e5431"
                    ]
                ],
                "1": [
                    "versions.yml:md5,ac2d0f2a8e77cd796d3b4b84fdd46b0d",
                    "versions.yml:md5,cb038a1f67a47a76d0e0a983cfd56789"
                ],
                "vcf_tbi": [
                    [
                        {
                            "id": "NA12878",
                            "panel": "1000GP",
                            "chr": "22",
                            "chunk": "16570065-16597215",
                            "tools": "quilt"
                        },
                        "NA12878_22_16570065-16597215.impute.annotate.vcf.gz:md5,f1a6f74a4782fc1df96db2b1fd458381",
                        "NA12878_22_16570065-16597215.impute.annotate.vcf.gz.tbi:md5,de3b20dc5f67aae2a9669f0c667edb5d"
                    ],
                    [
                        {
                            "id": "NA12878",
                            "panel": "1000GP",
                            "chr": "22",
                            "chunk": "16587172-16609999",
                            "tools": "quilt"
                        },
                        "NA12878_22_16587172-16609999.impute.annotate.vcf.gz:md5,d549aa4a26cbe659081002521c4400c2",
                        "NA12878_22_16587172-16609999.impute.annotate.vcf.gz.tbi:md5,674782f71cca76015e0b72f9bc4e5431"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,ac2d0f2a8e77cd796d3b4b84fdd46b0d",
                    "versions.yml:md5,cb038a1f67a47a76d0e0a983cfd56789"
                ]
            },
            [
                "VcfFile [chromosomes=[chr22], sampleCount=1, variantCount=572, phased=true, phasedAutodetect=false]",
                "VcfFile [chromosomes=[chr22], sampleCount=1, variantCount=570, phased=true, phasedAutodetect=false]"
            ],
            [
                [
                    "NA12878"
                ],
                [
                    "NA12878"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-11-06T16:24:29.617067286"
    }
}